sstables/kl/reader_impl.hh

```diff
@@ -170,277 +144,248 @@ private:
-                }
-                _ttl = _expiration = 0;
-                _deleted = (mask & column_mask::deletion) != column_mask::none;
-                _counter = false;
-                _state = state::CELL;
-            }
-            break;
+        co_return;
+    }
+    future<> produce_short_length_bytes(temporary_buffer<char>& buf) {
+        co_await read_nonempty();
+        data_consumer::read_status rs = read_short_length_bytes(_data, buf);
+        while (rs != data_consumer::read_status::ready) {
+            assert(_data.size() == 0);
+            co_await read_nonempty();
+            rs = primitive_consumer::consume(_data);
         }
-        case state::COUNTER_CELL:
-            if (read_64(data) != read_status::ready) {
-                _state = state::COUNTER_CELL_2;
-                break;
-            }
-            // fallthrough
-        case state::COUNTER_CELL_2:
-            // _timestamp_of_last_deletion = _u64;
-            _state = state::CELL;
-            goto state_CELL;
-        case state::EXPIRING_CELL:
-            if (read_32(data) != read_status::ready) {
-                _state = state::EXPIRING_CELL_2;
-                break;
-            }
-            // fallthrough
-        case state::EXPIRING_CELL_2:
-            _ttl = _u32;
-            if (read_32(data) != read_status::ready) {
-                _state = state::EXPIRING_CELL_3;
-                break;
-            }
-            // fallthrough
-        case state::EXPIRING_CELL_3:
-            _expiration = _u32;
-            _state = state::CELL;
-        state_CELL:
-        case state::CELL: {
-            if (read_64(data) != read_status::ready) {
-                _state = state::CELL_2;
-                break;
-            }
+        co_return;
+    }
+    future<> produce_bytes(fragmented_temporary_buffer& buf) {
+        co_await read_nonempty();
+        data_consumer::read_status rs = read_bytes(_data, _u32, buf);
+        while (rs != data_consumer::read_status::ready) {
+            assert(_data.size() == 0);
+            co_await read_nonempty();
+            rs = primitive_consumer::consume(_data);
         }
-        case state::CELL_2:
-            if (read_32(data) != read_status::ready) {
-                _state = state::CELL_VALUE_BYTES;
-                break;
-            }
-        case state::CELL_VALUE_BYTES:
-            if (read_bytes(data, _u32, _val_fragmented) != read_status::ready) {
-                _state = state::CELL_VALUE_BYTES_2;
-                break;
-            }
-        case state::CELL_VALUE_BYTES_2:
-        {
-            row_consumer::proceed ret;
-            if (_deleted) {
-                if (_val_fragmented.size_bytes() != 4) {
-                    throw malformed_sstable_exception("deleted cell expects local_deletion_time value");
+        co_return;
+    }
+    future<> wait() {
+        finish_reading();
+        _production_sem.signal();
```

Comment:
This is penalizing the common case where everything that needs to be read can be consumed in one, without ever yielding. This function will always go through the scheduler. perf_fast_forward (in the simplest case where it never fast-forwards) will report a huge increase in tasks and allocations here.

(let's also make it report instructions)